### PR TITLE
JDK 17 Official support

### LINKED ISSUE #1953: Update CI, tests & doc for Java 17 when GA
Java 17 (GA expected in September 2021) is an LTS release.

- [x] ensure infra repository has binaries for JDK 17
- [x] update CI integration tests to add JDK 17 to it https://github.com/elastic/apm-agent-java/pull/2184
- [x] review any test that is disabled on those JDKs, see usages of `org.junit.jupiter.api.condition.DisabledOnJre` (or the Junit4 equivalent).
- [x] update documentation for supported versions

### PR DESCRIPTION
**Pull request by SylvainJuge**
- remove few disabled tests on JDK 15,16 that are now working as expected thanks to #1942 
- update documentation
- remove jdk 12,13,14 from compatibility tests for now
- add jdk 18 to compatibility tests

## What does this PR do?

fixes #1953
supersedes #2158

### CODE REVIEW

File: Jenkinsfile [hunk 1]
```diff
@@ -333,7 +333,7 @@ pipeline {
           axis {
             // the list of support java versions can be found in the infra repo (ansible/roles/java/defaults/main.yml)
             name 'JAVA_VERSION'
-            values 'openjdk12', 'openjdk13', 'openjdk14', 'openjdk17'
+            values 'openjdk17', 'openjdk18'
```
Comment (jackshirazi) [inline Jenkinsfile:336]: is 18 available? I think it's pre-release

Comment (SylvainJuge) [reply Jenkinsfile:336]: Yes, it is available in the infra repo, but I haven't tested it so far, this PR will let us know for sure.

Comment (SylvainJuge) [reply Jenkinsfile:336]: Well, the good news is that OpenJDK 18 is available, the bad (and somehow expected) is that it fails.

For example Jacoco seems to fail due to this https://github.com/jacoco/jacoco/pull/1198, which is merged but not released, should be in version 0.8.8. Thus I'll disable it for now and create a separate issue to track all the tasks to support JDK 18.

Comment (SylvainJuge) [reply Jenkinsfile:336]: https://github.com/elastic/apm-agent-java/issues/2328 is the new shopping list for Java 18 compatibility when building the project.

**Comment by MatCuk** [comment]
Hi,

jdk17 is released now for quite some time. 
We migrated to jdk 17 and we are currently blocked by elastic apm as it's not working.
Is there any timeline when will this be released?

**Comment by SylvainJuge** [comment]
Hi @MatCuk , this PR is mostly about two things:
- making sure that we can build and test with JDK17 on CI, including disabling a few tests that are know to fail on JDK16 and JDK18
- document the support schedule

This PR is NOT about adding support to JDK17, which is already working as we don't have any major issue reported with it so far.

If you have any issue running the agent with JDK17 ? If so, I would suggest to:
- open a discussion in the forum here: https://discuss.elastic.co/c/apm.
- when (and if) the issue is confirmed open an issue with details and ideally steps to reproduce.

**Comment by MatCuk** [comment]
Ok I see. Thanks for the reply. 
I got some exception and when checking documentation I saw
https://www.elastic.co/guide/en/apm/agent/java/current/supported-technologies-details.html
only jdk11 is listed so I presumed this PR will add support. 
Might be some other issue then.

**Review by eyalkoren — APPROVED** [review]

**Comment by antemo** [comment]
Hi

Currently we are building our project with JDK17, but targeting JDK11 (until all apps can run with JDK17).

When running this app with JDK17 we get exception:
```
Failed to start agent
java.lang.reflect.InvocationTargetException
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:568)
        at co.elastic.apm.agent.bci.AgentMain.init(AgentMain.java:119)
        at co.elastic.apm.agent.bci.AgentMain.agentmain(AgentMain.java:69)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:568)
        at java.instrument/sun.instrument.InstrumentationImpl.loadClassAndStartAgent(InstrumentationImpl.java:491)
        at java.instrument/sun.instrument.InstrumentationImpl.loadClassAndCallAgentmain(InstrumentationImpl.java:513)
Caused by: java.lang.IllegalArgumentException: Unsupported class file major version 61
        at co.elastic.apm.agent.shaded.asm.ClassReader.<init>(ClassReader.java:196)
        at co.elastic.apm.agent.shaded.asm.ClassReader.<init>(ClassReader.java:177)
        at co.elastic.apm.agent.shaded.asm.ClassReader.<init>(ClassReader.java:163)
        at co.elastic.apm.agent.shaded.bytebuddy.utility.OpenedClassReader.of(OpenedClassReader.java:86)
        at co.elastic.apm.agent.shaded.bytebuddy.pool.TypePool$Default.parse(TypePool.java:679)
        at co.elastic.apm.agent.shaded.bytebuddy.pool.TypePool$Default.doDescribe(TypePool.java:665)
        at co.elastic.apm.agent.shaded.bytebuddy.pool.TypePool$Default$WithLazyResolution.access$001(TypePool.java:745)
        at co.elastic.apm.agent.shaded.bytebuddy.pool.TypePool$Default$WithLazyResolution.doResolve(TypePool.java:843)
        at co.elastic.apm.agent.shaded.bytebuddy.pool.TypePool$Default$WithLazyResolution$LazyResolution.isResolved(Type
Pool.java:872)
        at co.elastic.apm.agent.shaded.bytebuddy.pool.TypePool$Default$LazyTypeDescription$LazyAnnotationValue$ForTypeVa
lue.doResolve(TypePool.java:6058)
        at co.elastic.apm.agent.shaded.bytebuddy.pool.TypePool$Default$LazyTypeDescription$LazyAnnotationValue.filter(Ty
pePool.java:5995)
        at co.elastic.apm.agent.shaded.bytebuddy.description.annotation.AnnotationValue$AbstractBase.filter(AnnotationVa
lue.java:579)
        at co.elastic.apm.agent.shaded.bytebuddy.description.annotation.AnnotationDescription$AnnotationInvocationHandle
r.of(AnnotationDescription.java:253)
        at co.elastic.apm.agent.shaded.bytebuddy.pool.TypePool$Default$LazyTypeDescription$LazyAnnotationDescription$Loa
dable.load(TypePool.java:5964)
        at co.elastic.apm.agent.bci.ElasticApmAgent.validateAdvice(ElasticApmAgent.java:481)
        at co.elastic.apm.agent.bci.ElasticApmAgent.getTransformer(ElasticApmAgent.java:428)
        at co.elastic.apm.agent.bci.ElasticApmAgent.applyAdvice(ElasticApmAgent.java:390)
        at co.elastic.apm.agent.bci.ElasticApmAgent.initAgentBuilder(ElasticApmAgent.java:313)
        at co.elastic.apm.agent.bci.ElasticApmAgent.initInstrumentation(ElasticApmAgent.java:260)
        at co.elastic.apm.agent.bci.ElasticApmAgent.initInstrumentation(ElasticApmAgent.java:164)
        at co.elastic.apm.agent.bci.ElasticApmAgent.initialize(ElasticApmAgent.java:150)
        ... 12 more
```
When running same app with JDK11 we get no issues.

Is this a known problem? And if so, is there any way to resolve this?

**Comment by SylvainJuge** [comment]
Hi @antemo ,

This is not expected, which version of the agent are you using here ? This error indicates an old version of the agent is used as there is still class shading used for agent dependencies thus it's pre `1.28.x` (now bytebuddy and asm dependencies remain in their original packages).

Could you try with latest agent version `1.28.4` to see if the issue persists ? If yes then opening a new issue might be relevant here.

**Pull request merged**